src/schemaValidation/validateSchema.ts

```diff
@@ -0,0 +1,169 @@
+import { Manifest } from "../releaseFiles/manifest/types";
+import { Compose } from "../releaseFiles/compose/types";
+
+const ajv = new Ajv({
+  allErrors: true,
+  coerceTypes: true,
+  strictSchema: false
+});
+
+ajvErrors(ajv);
+
+console.log("Validating manifest schema");
+const manifestValidate = ajv.compile(manifestSchema);
+console.log("Validating setupwizard schema");
+const setupWizardValidate = ajv.compile(setupWizardSchema);
+console.log("Validating compose schema");
+const composeValidate = ajv.compile(composeSchema); // compose schema https://raw.githubusercontent.com/compose-spec/compose-spec/master/schema/compose-spec.json
+
+/**
+ * Validates a manifest syncronously. Does NOT throw.
+ * @param manifest
+ * @returns = {
+ *   valid: false|true
+ *   errors: [
+ *     "manifest should have required property 'description'",
+ *     "manifest.image.path should be an non-empty string",
+ *   ]
+ * }
+ */
+export function validateSchema(
+  releaseFile: Manifest | Compose | SetupWizard
+): void {
+  const { valid, errors } = validateReleaseFile(releaseFile);
+  if (valid) return;
+
+  // If not valid, print errors and stop execution
+  throw new CliError(
+    `Invalid ${releaseFile}: \n${errors.map(msg => `  - ${msg}`).join("\n")}`
+  );
+}
+
+/**
+ * Validates a manifest syncronously. Does NOT throw.
+ * @param manifest
+ * @returns = {
+ *   valid: false|true
+ *   errors: [
+ *     "manifest should have required property 'description'",
+ *     "manifest.image.path should be an non-empty string",
+ *   ]
+ * }
+ */
+export function validateReleaseFile(
+  releaseFile: Manifest | Compose | SetupWizard
+): { valid: boolean; errors: string[] } {
+  const releaseFileType = inferReleaseFileType(releaseFile);
+  switch (releaseFileType) {
+    case "manifest":
+      return {
+        valid: Boolean(manifestValidate(releaseFile)),
+        errors: manifestValidate.errors
+          ? manifestValidate.errors.map(e => processError(e, releaseFileType))
+          : []
+      };
+    case "compose":
+      return {
+        valid: Boolean(composeValidate(releaseFile)),
+        errors: composeValidate.errors
+          ? composeValidate.errors.map(e => processError(e, releaseFileType))
+          : []
+      };
+    case "setupWizard":
+      return {
+        valid: Boolean(setupWizardValidate(releaseFile)),
+        errors: setupWizardValidate.errors
+          ? setupWizardValidate.errors.map(e =>
+              processError(e, releaseFileType)
+            )
+          : []
+      };
+    default:
+      throw new Error(`Unknown release file type: ${releaseFileType}`);
+  }
+}
+
+function inferReleaseFileType(
+  releaseFile: Manifest | Compose | SetupWizard
+): ReleaseFileType {
+  if ("name" in releaseFile) {
+    return "manifest";
+  } else if ("services" in releaseFile) {
+    return "compose";
+  } else if ("fields" in releaseFile) {
+    return "setupWizard";
+  } else throw Error("Unknown release file type");
```

Comment:
This function is super dangerous, and not worth the trouble. There may be key collisions in the future

Reply:
removed